x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/ShapeDocValuesQuery.java

```diff
@@ -0,0 +1,197 @@
+            sb.append(relation);
+            sb.append(':');
+        }
+        sb.append(Arrays.toString(geometries));
+        return sb.toString();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (sameClassAs(obj) == false) {
+            return false;
+        }
+        ShapeDocValuesQuery<?> other = (ShapeDocValuesQuery<?>) obj;
+        return field.equals(other.field) && relation == other.relation && Arrays.equals(geometries, other.geometries);
+    }
+
+    @Override
+    public int hashCode() {
+        int h = classHash();
+        h = 31 * h + field.hashCode();
+        h = 31 * h + relation.hashCode();
+        h = 31 * h + Arrays.hashCode(geometries);
+        return h;
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        if (visitor.acceptField(field)) {
+            visitor.visitLeaf(this);
+        }
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
+        if (relation == ShapeField.QueryRelation.CONTAINS) {
+            return getContainsWeight(scoreMode, boost);
+        } else {
+            return getStandardWeight(scoreMode, boost);
+        }
+    }
+
+    private ConstantScoreWeight getStandardWeight(ScoreMode scoreMode, float boost) {
+        return new ConstantScoreWeight(this, boost) {
+            final Component2D component2D = create(geometries);
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                final BinaryDocValues values = context.reader().getBinaryDocValues(field);
+                if (values == null) {
+                    return null;
+                }
+                final GeometryDocValueReader reader = new GeometryDocValueReader();
+                final Component2DVisitor visitor = Component2DVisitor.getVisitor(component2D, relation, encoder);
+
+                final TwoPhaseIterator iterator = new TwoPhaseIterator(values) {
+
+                    @Override
+                    public boolean matches() throws IOException {
+                        reader.reset(values.binaryValue());
+                        visitor.reset();
+                        reader.visit(visitor);
+                        return visitor.matches();
+                    }
+
+                    @Override
+                    public float matchCost() {
+                        return 1000f; // TODO: what should it be?
+                    }
+                };
+                return new ConstantScoreScorer(this, boost, scoreMode, iterator);
+            }
+
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                return DocValues.isCacheable(ctx, field);
+            }
+
+        };
+    }
+
+    /** Override if special cases, like dateline support, should be considered */
+    protected boolean addSpecialCase(List<Component2D> components2D, GEOMETRY geometry) {
+        return false;
```

Comment:
Not strong opinion but I wonder if we should change the API to somethng like:

```
    protected abstract Component2D create(GEOMETRY... geometry);

    protected abstract void add(List<Component2D> components2D, GEOMETRY geometry);
```

So we don't need this special case.

Reply:
Interesting approach. Is easier to read and reason about, so I've changed it to that, sort-of.